yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala

```diff
@@ -82,6 +102,180 @@ class YarnSparkHadoopUtil extends SparkHadoopUtil {
     if (credentials != null) credentials.getSecretKey(new Text(key)) else null
   }
 
+  private[spark] override def scheduleLoginFromKeytab(): Unit = {
+    val principal = System.getenv("SPARK_PRINCIPAL")
+    val keytab = System.getenv("SPARK_KEYTAB")
+    if (principal != null) {
+      val delegationTokenRenewerThread =
+        new Runnable {
+          override def run(): Unit = {
+            if (!loggedInViaKeytab) {
+              // Keytab is copied by YARN to the working directory of the AM, so full path is
+              // not needed.
+              loggedInUGI = UserGroupInformation.loginUserFromKeytabAndReturnUGI(
+                principal, keytab)
+              loggedInViaKeytab = true
+            }
+            val nns = getNameNodesToAccess(sparkConf)
+            val newCredentials = loggedInUGI.getCredentials
+            obtainTokensForNamenodes(nns, conf, newCredentials)
+            val remoteFs = FileSystem.get(conf)
+            val stagingDirPath =
+              new Path(remoteFs.getHomeDirectory, System.getenv("SPARK_YARN_STAGING_DIR"))
+            val tokenPathStr = sparkConf.get("spark.yarn.credentials.file")
```

Comment:
Couldn't you set `spark.yarn.credentials.file` to the full URL and avoid having to look at `SPARK_YARN_STAGING_DIR` in multiple places?